### ISSUE TITLE
 8-bit colors not supported 

### ISSUE THREAD

**Issue by 1268** [gwsw/less#410]
<https://wikipedia.org/wiki/ANSI_escape_code#8-bit>

using this Go program:

~~~go
package main

func main() {
   print("\x1b[38;5;2m+asdf\x1b[39m\n")
   print("\x1b[32m+\x1b[m\x1b[32masdf\x1b[m\n")
}
~~~

I get expected result:

![pass](https://github.com/jftuga/less-Windows/assets/126826528/0fb3b53c-5ceb-45b1-9006-091efb21a8ce)

if I write the same data to a file:

~~~go
package main

import "os"

func main() {
   f, err := os.Create("hello.txt")
   if err != nil {
      panic(err)
   }
   defer f.Close()
   f.WriteString("\x1b[38;5;2m+asdf\x1b[39m\n")
   f.WriteString("\x1b[32m+\x1b[m\x1b[32masdf\x1b[m\n")
}
~~~

Less fails with the first line:

![fail](https://github.com/jftuga/less-Windows/assets/126826528/41944cbf-e30f-484b-abd4-665fc8077bba)

~~~
> less --version
less 633 (Spencer V8 regular expressions)
~~~

**Comment by 1268** [comment]
https://github.com/martinvonz/jj/issues/1930

**Comment by polluks** [comment]
Windows-Terminal 1.18.1462.0 works fine
![2023-08-02 11_48_35-Clipboard](https://github.com/gwsw/less/assets/4707573/def92bd0-80ce-45d9-b30c-cdc505674213)

**Comment by 1268** [comment]
that is not PowerShell or cmd.exe. I am not interested in the results of third party shells.

**Comment by 1268** [comment]
https://github.com/gwsw/less/issues/377

**Comment by zadjii-msft** [comment]
This looks like it's a specific issue with the Windows port of `less`

![image](https://github.com/gwsw/less/assets/18356694/a2a25f64-88bd-49d1-92db-22063ff2a2bc)

**Comment by 1268** [comment]
I have now blocked @polluks @zadjii-msft.

I would ask any other users to respect that this is a Windows issue, so please do any testing with first party Windows tools only, AKA PowerShell or cmd.exe

**Comment by gwsw** [comment]
Even if an issue only appears on Windows, it is perfectly reasonable and in fact **helpful** to compare the behavior to other systems and other terminals, to help isolate where the bug manifests. Even if you personally don't care about non-Windows systems or third party terminals, it is completely unreasonable and counterproductive to attempt to prohibit discussion of such environments. Please do not do that again.

Regarding this issue, 8-bit color sequences are not supported on Windows because less uses ```_settextcolor``` and ```_setbkcolor```, which only support 4-bit color. One solution would be to use -Da which passes SGR sequences to the terminal, if the terminal supports ENABLE_VIRTUAL_TERMINAL_PROCESSING. I don't know how to use 8-bit color on a non-virtual terminal or even if it's possible, but I will look into it after #383 is merged.

**Comment by DHowett** [comment]
> I would ask any other users to respect that this is a Windows issue, so please do any testing with first party Windows tools only

You are welcome to block whomever you please, of course, but I would be remiss if I didn't mention:

1. Windows Terminal is a first-party Windows tool, built by Windows engineers and checked into Windows.
2. zadjii-msft is one of the engineers on the Windows Console subsystem, which is a first-party Windows subsystem built by Windows engineers and checked into Windows.

As far as finding somebody to talk about the Windows side of a Windows bug with, he's one of the best resources available... on account of it is the console subsystem (and the Terminal product, which is--again--owned by the same team) that is involved in a lot of the VT parsing and all of the console mode handling that Windows applications use.

> Please do not do that again.

(Thank you, @gwsw. I was writing this up and saw your comment appear. Full agree on all points. :smile:)

**Comment by 1268** [comment]
@DHowett but Bash is not a Microsoft product, that's the point that seems to be getting repeated missed here. Bash is different enough that we cant even make assumptions about the Less being used in that case. a Windows Bash user could be running a MSYS2 or even Cygwin variant of Less, which would be pointless in regard to troubleshooting the issue as posted, which involves NATIVE WINDOWS tools.

**Comment by gwsw** [comment]
I don't understand why you are mentioning bash. No one has mentioned bash in this thread so far before you did. But the shell is irrelevant; the shell does not interpret SGR sequences, the terminal does that. So you could rightfully object to someone comparing the Windows terminal to xterm or uxrvt, but bash is irrelevant.

I won't close this issue because it is a valid concern, but I'd appreciate it if you don't post any more here. You've made your point, but this further belligerent argumentation is not helping anyone.

**Comment by 1268** [comment]
> I don't understand why you are mentioning bash. No one has mentioned bash in this thread so far before you did.

every single picture posted so far is Bash, except mine. this is important.

> But the shell is irrelevant;

its not. If someone is using Bash on Windows, then likely they are using one of these:

- MSYS2
- Cygwin
- WSL

regardless of which, in that situation is highly likely that they are running a LINUX version of Less. either pure Linux in the case of WSL, or by using the Linux emulation layer DLL via MSYS2 or Cygwin. in all of those cases, its going to skew any results, because in all likelihood THEY AREN'T using a native Windows Less. So yes, the shell is quite important in the context of this issue.

> I won't close this issue because it is a valid concern, but I'd appreciate it if you don't post any more here. You've made your point, but this further belligerent argumentation is not helping anyone.

Again I will not stop. I am an expert on the subject matter, in regards to Windows/MSYS2/Cygwin terminal use and similar. So I have a place here. if you want to ignore my valid and reasoned objections because you dont like my tone, thats your choice. but I would advise against making an emotional decision, and instead focus on the facts of the issue as I am trying to provide them, and lets see about solving the issue rather than bickering.

**Comment by gwsw** [comment]
I've already explained what the issue is, 5 posts above this one. Perhaps you overlooked it in your desire to control the conversation. If you have knowledge of terminal use and similar, and wanted to help resolve this, you could have suggested ways to use 8-bit color on a non-virtual terminal rather than carping about who is allowed to say what in YOUR ISSUE. 

Anyway, I am closing this issue which you have derailed, and will open a new one later. You are not welcome to participate in that conversation.

**Comment by 1268** [comment]
when I posted this:

> see about solving the issue rather than bickering

I see you weren't able to do that:

> Perhaps you overlooked it in your desire to control the conversation.

> allowed to say what in YOUR ISSUE.

> I am closing this issue which you have derailed, 

I asked that you focus on the facts of the case, which I repeatedly provided. Using Windows with Bash shell is likely going to skew any result. as such, I blocked people doing that, in order to prevent any lines of conversation like that. however some people cant handle being blocked or even seeing others blocked, even though their comments are wholly off topic in regards to the issue. If people want to be distracted by my attempts to keep the conversation on topic as you have, then thats a shame.
